src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -363,8 +431,316 @@ private static unsafe nuint GetIndexOfFirstNonAsciiByte_Sse2(byte* pBuffer, nuin
             // Tzcnt is the correct operation to count the number of zero bits quickly. If this instruction isn't
             // available, we'll fall back to a normal loop.
 
-            Debug.Assert(currentMask != 0, "Shouldn't be here unless we see non-ASCII data.");
-            pBuffer += (uint)BitOperations.TrailingZeroCount(currentMask);
+            Debug.Assert(ContainsNonAsciiByte_Sse2(currentMask), "Shouldn't be here unless we see non-ASCII data.");
+            if (Sse2.IsSupported)
+            {
+                pBuffer += (uint)BitOperations.TrailingZeroCount(currentMask);
+            }
+            else
+            {
+                throw new PlatformNotSupportedException();
+            }
+
+            goto Finish;
+
+        FoundNonAsciiDataInCurrentDWord:
+
+            uint currentDWord;
+            Debug.Assert(!AllBytesInUInt32AreAscii(currentDWord), "Shouldn't be here unless we see non-ASCII data.");
+            pBuffer += CountNumberOfLeadingAsciiBytesFromUInt32WithSomeNonAsciiData(currentDWord);
+
+            goto Finish;
+
+        InputBufferLessThanOneVectorInLength:
```

Comment:
>InputBufferLessThanOneVectorInLength [](start = 8, length = 36)

I get it that you will have to special case `SSE2` and `AdvSimd` throughout this method so you separated those 2 methods, but now we have this code block under `InputBufferLessThanOneVectorInLength` which doesn't have any special handing for SSE2/AdvSimd but is repeated in 2 places. In future, if we need to make changes/bug fix in this, it will be easy to forget to update the other copy. Same applies to other code inside this method. I would recommend having a single copy with if/else checks for SSE2/AdvSimd.

Reply:
No actually, the special casing wasn't my main concern. The main concern was that the compiler would complain about using uninitialized `currentSseMask`. Even though we initialize it before using it inside the `if Sse2` block (and we use it only inside `if Sse2` blocks), that information won't flow to the compiler fully. And because `currentSseMask` is a `uint`, there's no straightforward sentinel value I can initialize`currentSseMask` to. Both `uint.MinValue` and `uint.MaxValue` are valid values :)

Reply:
One potential fix is to the use a single variable `currentSseMaskOrAdvSimdIndex` for both the `Sse2` and `AdvSimd` cases and carefully call the appropriate `ContainsNonAsciiByte_Sse2` or `ContainsNonAsciiByte_AdvSimd` method inside the `if` blocks

Reply:
> Both uint.MinValue and uint.MaxValue are valid values :)

Hhm, I don't think `uint.MaxValue` can be a valid value. For SSE2, max value you can get is 65535. For AdvSimd, it can be `16`, I don't remember why the assert is `Debug.Assert((mask != 0) ? index < 16 : index >= 16);` has `index >= 16`. I might have to try out some inputs to know for sure. But regardless, I don't think it can ever be `uint.MaxValue` so you can just initialize `currentMask` and `currentIndex` to that value and use if-then.

Reply:
Oh yea you're right. Fixed now, changed them to different variables and added Asserts in `ContainsNonAsciiByte` to make sure they are assigned before use.